@ekino/logger v3.0

tduyng opened the pull request:
## Summary
Rewrite @ekino/logger for v3.0

## Motivation
- It's been so long time the library is not matained
- Rewrite this library for adapting the modern Node.js (Typescript, ESM)

## What has changed?
Here are the key changes made in this PR:
- Rewritten in ESM
- Support for dual package (CommonJS and ESM)
- Rewritten in TypeScript
- Removed many unnecessary dependencies
- Replaced yarn with pnpm, which is faster and more convenient with different commands and management
- Upgraded GitHub Actions, rewrote some deprecated syntax
- Using git-cliff on GitHub Actions for easily customizable CHANGELOG
- Replaced ESLint + Prettier with Biome, which is faster, has sufficient rules, and is easier to set up
- Rewrote all tests from ava + sinon.js + nyc + ts-node to vitest (compatible with ESM)
- Create a benchmark compare with pino.js and winston
- Refactored and cleaned up code, focusing on performance, and much more...

## Test plan
- All unit tests have been correctly rewritten with Vitest
- All functional test behaviors are ensured

Also close issue #25 , close issue #37

tduyng committed:
chore(yarn): use yarn v2

tduyng committed:
chore(hooks): remove husky, use native githooks

tduyng committed:
chore(deps): use exact version of dev dependencies for better local dev

tduyng committed:
chore(lint): implement biome and remove eslint,prettier

tduyng committed:
fix(lint): fix all linter issues

tduyng committed:
refactor(outils): remove unused function

tduyng committed:
refactor: write native colors outils

tduyng committed:
refactor(examples): give meaningfull messages

tduyng committed:
chore(deps): remove unused eslint and prettier

tduyng committed:
refactor(adapters): create wrapper writable stream for process.stdout.write

tduyng committed:
feat(perf): implement fast-json-stringify for better performance json parser

tduyng committed:
refactor: implement pur functional, avoid mutable internal object

tduyng committed:
refactor: separate default config and mutual config

tduyng committed:
fix: handle correctly dual signature for logMethod

tduyng committed:
feat(perf): implement memorization for caching expensive operations

tduyng committed:
refactor: separate json schema for fast-json-stringify

tduyng committed:
fix: fast-json-stringify do not support replacer

tduyng committed:
feat(3.0): convert src code to ESM

tduyng committed:
feat(3.0): remove ts-node, use tsx, convert examples to typescript

tduyng committed:
chore(3.0): ignore format package.json by biome

tduyng committed:
feat(3.0): support more colors

tduyng committed:
chore(3.0): support dual npm package in ESM and CJS

tduyng committed:
chore(3.0): set Node.js 18 as the minimum supported version

tduyng committed:
chore(3.0): remove unused Makefile

tduyng committed:
ci(3.0): fix, update action versions, run test on Nodejs18,20,22

tduyng committed:
ci(3.0): simplify release step and build Typescript files

tduyng committed:
chore(3.0): run examples/*.ts with deno for faster execution

tduyng committed:
chore(3.0): remove unused dependencies

tduyng committed:
chore(3.0): add more tags

tduyng committed:
ci(3.0): add issue templates: bug, feature_request

tduyng committed:
docs(3.0): update image examples

tduyng committed:
test(3.0): rewrite tests with vitest for ESM

tduyng committed:
test(3.0): reimplement coverage with vitest + c8

tduyng committed:
feat(3.0): use native crypto.randomUUID to replace uuidV4

tduyng committed:
test(3.0): rewrite all test files in Typescript with vitest

tduyng committed:
chore(3.0): cleanup package.json

tduyng committed:
docs(3.0): update documentations

tduyng committed:
chore(3.0): add CONTRIBUTION rules

tduyng committed:
docs(3.0): add breaking change docs

tduyng committed:
ci(3.0): add type checking step

tduyng committed:
docs(3.0): create benchmark with v2, winston and pino

tduyng committed:
perf(3.0): early return log when not forcing and not active level

tduyng committed:
perf(3.0): avoid using Object.assign

tduyng committed:
docs(3.0): add benchmarks results

tduyng committed:
chore(3.0): add benchmark for complex objects

tduyng committed:
chore(3.0): simplify and split and create helpers functions for benchmarking

tduyng committed:
chore(3.0): remove tsx, write build script directly in js

tduyng committed:
chore(3.0): off useImportType lint rule

tduyng committed:
test(3.0): use correct type MockInstance for vi.spyOn

tduyng committed:
ci(3.0): implement git cliff for highly customizable changelog

tduyng committed:
chore(3.0): update the dev dependencies

tduyng committed:
chore(3.0): replace yarn by pnpm

tduyng committed:
chore(3.0): upgrade dev dependencies

tduyng committed:
feat(3.0): remove fast-json-stringify

tduyng committed:
chore(3.0): simplify build scripts

tduyng committed:
chore(deps): update latest dependencies

tduyng committed:
ci: setup bot user for creating release commit

tduyng committed:
ci: simplify bump version and tag with npm

tduyng committed:
feat: use @tduyng/prettyouput to replace deprecated prettyoutput library

tduyng committed:
docs: update README.md

tduyng committed:
chore(deps): use major version for dependencies field

tduyng committed:
ci: check typings for all files

tduyng commented:
TODO: fix github-action for release and changelog.

tduyng committed:
feat: remove support for node18, support test for node24

tduyng committed:
chore: update dev dependencies

tduyng committed:
chore: move prettyoutput to optional dependencies, not always needed

Nico385412 commented on CONTRIBUTING.md:
```diff
@@ -0,0 +1,77 @@
+
+## How to Contribute
+
+Follow this guide to ensure a smooth contribution process. If you're planning significant work or substantial changes, please create an issue labeled "contribution enquiry" to discuss your proposal. This will help ensure that your work aligns with the project's goals.
```
1. The label does not exist
2. A new contributor cannot add labels, most of the time a member must set the labels for him

tduyng replied:
Yes, I simplify that. Thank Nico.

Nico385412 commented on CONTRIBUTING.md:
````diff
@@ -0,0 +1,77 @@
+2. **Clone Your Fork**  
+   Clone your forked repo to your local machine:
+   ```bash
+   git clone git@github.com:ekino/node-logger.git
````
```suggestion
   git clone git@github.com:[YOUR_GITHUB_ACCOUNT]/node-logger.git
```

since you forked the repository the address is not correct

tduyng replied:
You are right! Done!

Nico385412 commented on CONTRIBUTING.md:
```diff
@@ -0,0 +1,77 @@
+   Ensure your changes are reliable by including tests. For small patches, a simple test is fine; for new features, create a dedicated test suite.
+
+7. **Coding Style and Conventions**  
+   Maintain the project’s coding standards. Consistent indentation, comments, and style improve readability.
```
maybe we could invite the contributor to install official biome ext for vscode ? (or any other IDE)
https://biomejs.dev/reference/vscode/

tduyng replied:
Yes, that's should be interesting point. Done!

Nico385412 commented on benchmarks/README.md:
````diff
@@ -0,0 +1,36 @@
+
+- Benchmark basic
+```bash
+node ./benchmarks/basic.js
````
```suggestion
pnpm run build
node ./benchmarks/basic.js
```

tduyng replied:
Done!

Nico385412 commented on benchmarks/README.md:
````diff
@@ -0,0 +1,36 @@
+- Benchmark with complex object
+
+```bash
+node ./benchmarks/complex.js
````
```suggestion
pnpm run build
node ./benchmarks/complex.js
```

tduyng replied:
Done!

Nico385412 commented on benchmarks/README.md:
````diff
@@ -0,0 +1,36 @@
+│ 3       │ 'Winston'            │ '96,834'  │
+└─────────┴──────────────────────┴───────────┘
+```
+
````
```suggestion
Tested on Nodejs 24.3.0 
Macbook Pro AppleM2 Pro (macos 15.7.2)
```bash
┌─────────┬──────────────────────┬───────────┐
│ (index) │ library              │ ops/sec   │
├─────────┼──────────────────────┼───────────┤
│ 0       │ 'Pino'               │ '347,826' │
│ 1       │ '@ekino/logger v3.x' │ '333,222' │
│ 2       │ '@ekino/logger v2.x' │ '311,624' │
│ 3       │ 'Winston'            │ '16,353'  │
└─────────┴──────────────────────┴───────────┘
```

```

tduyng replied:
Yes, I updated that with apple M4 and latest node 24. Thank you for the suggestion.

Nico385412 commented on benchmarks/utils.js:
```diff
@@ -0,0 +1,97 @@
+ * @param {Suite} suite Benchmark suite instance
+ * @param {number} timeout Timeout value in milliseconds
+ */
+export const runSuiteWithDynamicTimeout = async (suite, timeout = 1000) => {
+    const startTime = performance.now()
+
+    const results = await suite.run()
+
+    const endTime = performance.now()
+    const executionTime = endTime - startTime
+    const timeoutValue = executionTime + timeout
+
+    const table = formatResultsTable(results)
+
+    setTimeout(() => console.table(table), timeoutValue)
+}
```
```suggestion
```

Nico385412 commented on benchmarks/basic.js:
```diff
@@ -0,0 +1,47 @@
+        })
+    })
+
+await runSuiteWithDynamicTimeout(suite)
```
```suggestion
const table = await runSuiteWithDynamicTimeout(suite)

process.on("exit", () => {
    console.table(table)
})
```

tduyng replied:
Done!

Nico385412 commented on benchmarks/complex.js:
```diff
@@ -0,0 +1,41 @@
+        })
+    })
+
+await runSuiteWithDynamicTimeout(suite, 40000)
```
```suggestion
const table = await runSuiteWithDynamicTimeout(suite)

process.on("exit", () => {
    console.table(table)
})
```

tduyng replied:
Added!

Nico385412 commented on benchmarks/README.md:
````diff
@@ -0,0 +1,36 @@
+│ 2       │ '@ekino/logger v2.x' │ '18,826' │
+│ 3       │ 'Winston'            │ '4,251'  │
+└─────────┴──────────────────────┴──────────┘
+```
````
```suggestion
Tested on Nodejs 24.3.0
Macbook Pro AppleM2 Pro (macos 15.7.2)
```bash
┌─────────┬──────────────────────┬──────────┐
│ (index) │ library              │ ops/sec  │
├─────────┼──────────────────────┼──────────┤
│ 0       │ 'Pino'               │ '42,089' │
│ 1       │ '@ekino/logger v3.x' │ '31,158' │
│ 2       │ '@ekino/logger v2.x' │ '26,871' │
│ 3       │ 'Winston'            │ '2,743'  │
└─────────┴──────────────────────┴──────────┘
```
```

tduyng replied:
Updated!

Nico385412 commented on docs/BREAKING_CHANGES.md:
```diff
@@ -0,0 +1,12 @@
+
+- **Dual ESM and CommonJS**: Added support for both module formats. No syntax changes needed from 2.x.
+- **Functional style**: Refactored to use a fully functional API, removing internal objects. Update code that accessed internals directly.
+- **Node.js update**: Dropped support for Node 14 and 16. Now compatible with Node 18, 20, 22, and above.
```
```suggestion
- **Node.js update**: Dropped support for Node 16, and 18. Now compatible with Node 20, 22, 24, and above.
```

tduyng replied:
It’s been a long time since I made this PR :)) Thanks, Nico, I’ve updated it.

Nico385412 commented on examples/cjs/example3.cjs:
```diff
@@ -0,0 +1,11 @@
+
+logger.setNamespaces('namespace:*')
+logger.setLevel('debug')
+//logger.setOutput(logger.outputs.json)
```
```suggestion
```

tduyng replied:
Done!

Nico385412 commented on examples/esm/example3.js:
```diff
@@ -0,0 +1,11 @@
+
+logger.setNamespaces('namespace:*')
+logger.setLevel('debug')
+//logger.setOutput(logger.outputs.json)
```
```suggestion
```

tduyng replied:
Done!

Nico385412 commented on examples/example3.ts:
```diff
@@ -1,8 +1,11 @@
 
 logger.setNamespaces('namespace:*')
 logger.setLevel('debug')
 //logger.setOutput(logger.outputs.json)
```
```suggestion
```

tduyng replied:
Done!

tduyng committed:
docs: correct informations

tduyng committed:
chore: update benchmarks

tduyng committed:
chore: remove dead comments and update breaking changes for node version

tduyng committed:
ci: run only release by maintainer or admin

tduyng commented:
Thank @Nico385412 for all the reviews!

tduyng committed:
test: improve coverages and show coverage summary

atacanjacques commented on package.json:
```diff
@@ -23,61 +23,56 @@
+        "winston": "3.18.3"
+    },
+    "optionalDependencies": {
+        "@tduyng/prettyoutput": "2.x"
```
Why put this dependancy in optionalDependencies ?

tduyng replied:
The pretty feature is optional and mainly intended for local development and easier debugging.
In production, logs are emitted in JSON format.

The pretty formatter is therefore declared as an optional dependency. Users are free to use @tduyng/prettyoutput or provide their own implementation, or disable pretty logging entirely. This keeps the setup flexible and avoids imposing a specific formatter.

atacanjacques commented on package.json:
```diff
@@ -23,61 +23,56 @@
+        "examples-esm": "./examples/esm/exec.sh",
+        "examples-cjs": "./examples/cjs/exec.sh",
+        "benchmark": "node ./benchmarks/basic.js",
+        "postinstall": "git config core.hooksPath ./.githooks || true"
```
Since this modify the user git config for the repo, maybe we should add this script as a step on the readme instead of running it automatically ?
What do you think ?

tduyng replied:
Yes, added. Thank for your suggestion @atacanjacques

halimatou-cpu reviewed (approved):
👌

tduyng committed:
chore: rename postinstall to hook script

tduyng committed:
feat: use native node util.inspect for pretty json

tduyng committed:
feat: update dependencies and typescript 6

tduyng committed:
ci: implement OICD npm release workflow

tduyng committed:
fix: resolve regressions and remove performance benchmarks

- Fix error message format for backward compatibility
- Fix setNamespaces whitespace handling
- Fix setGlobalContext to use deep copy (structuredClone)
- Export isEnabled function for namespace/level checking
- Add cache invalidation when namespaces/level change
- Remove memoize to avoid memory leak
- Remove benchmarks and performance docs (not needed)
- Update BREAKING_CHANGES.md

tduyng committed:
chore: stop supporting node 20 (EOL)
